front/src/modules/trainschedule/components/Timetable/Timetable.tsx

```diff
@@ -104,17 +117,82 @@ export default function Timetable({
     return true;
   };
 
+  const specialCodeDictionary: { [key: string]: string } = {
+    'Divers - Haut le pied': 'HLP',
+    '': 'NO CODE',
+  };
+
+  const extractTagCode = (tag: string | null) => {
+    if (!tag) {
+      return 'NO CODE';
+    }
+    if (tag in specialCodeDictionary) {
+      return specialCodeDictionary[tag];
+    }
+
+    const regex = /\s*(\w+)\s*$/;
+    const matches = tag.match(regex);
+    if (matches && matches[1]) {
+      return matches[1];
+    }
+
+    return tag;
+  };
+
   const trainsList = useMemo(() => {
-    if (timetable) {
-      return timetable.train_schedule_summaries
-        .filter((train) => keepTrain(train, debouncedTerm))
-        .map((train) => ({
+    if (!timetable) return [];
+
+    return timetable.train_schedule_summaries
+      .filter((train) => {
+        if (!keepTrain(train, debouncedFilter)) return false;
+
+        // Apply validity filter
+        if (validityFilter === 'valid' && train.invalid_reasons?.length > 0) return false;
+        if (
+          validityFilter === 'invalid' &&
+          (!train.invalid_reasons || train.invalid_reasons.length === 0)
+        )
+          return false;
+
+        // Apply tag filter
+        if (selectedTags.size > 0 && !selectedTags.has(extractTagCode(train.speed_limit_tags)))
+          return false;
+
+        // Apply rolling stock filter
+        const rollingStock = rollingStocks.find((rs) => rs.id === train.rolling_stock_id);
+        const rollingStockDetail = rollingStock?.metadata?.detail || '';
+        const rollingStockSeries = rollingStock?.metadata?.series || '';
+        const rollingStockFamily = rollingStock?.metadata?.family || '';
+        const rollingStockName = rollingStock?.name || '';
```

Comment:
```suggestion
        const { family = '', detail = '', series = '' } = rollingStock?.metadata || {};
```
You don't need to check the name of the rs, it's already done by the `keepTrain()` function l.147

Reply:
Done